### ISSUE TITLE
`model.fit()` ends before all epochs are ran

### ISSUE THREAD

**Issue by HungWoeiNeoh** [tensorflow/tensorflow#47938]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04 (in a Docker container)
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.4.1
- Python version: 3.6.9
- CUDA/cuDNN version: CUDA 11.0 with cuDNN 8.0.5.39
- GPU model and memory: GeForce RTX 2080 Ti with 11019 MiB

**Describe the current behavior**
I have been trying to implement [FlowNet 2.0](https://arxiv.org/abs/1612.01925) using Tensorflow 2.0 as an exercise. I've managed to implement a single FlowNetS module, but have been unable to train it. My `model.fit()` function is ran with 480 epochs, with 2500 steps per epoch. For some reason, `model.fit()` always ends after 2 epochs.

**Describe the expected behavior**
The code runs for all 480 epochs (this will take, by my estimation, 200 days as the dataset loading is slow).

**Standalone code to reproduce the issue**
This issue probably cannot be reproduced in Colab due to the large dataset and long runtime.

Here's some details of my implementation:
- subclassed from `tf.keras.Model`, which has a custom implementation of `__init__()`, `call()` and `test_step()`. The only thing I changed in `test_step()` is the metrics that get returned by `test_step()`
- A custom loss class and a custom metric class
- Adam optimizer with PiecewiseConstantDecay schedule
- A `tf.data.Dataset` that dynamically loads the images and flows as necessary, and passes the data to the network in batches of 8, with prefetching. I cannot load the full Flying Chairs dataset into memory, since it is 30 GB
- Learning done with the standard calls of `model.compile()` and `model.fit()`. The model returns multiple outputs during training, so loss_weights are provided during compilation. The `fit()` command is as below, with 2 `ModelCheckpoint` callbacks to store the full model of the best and last checkpoint (using this [temporary workaround](https://github.com/tensorflow/tensorflow/issues/42741#issuecomment-706534711)):
```
history = net.fit(x=train_set, epochs=480, verbose=1, validation_data=val_set,
                      steps_per_epoch=2500, callbacks=[best_checkpoint, last_checkpoint])
```

I'm not sure how I'm supposed to include the source code, since the project source code is split into multiple files across subdirectories. Should I compile all the relevant code into a single source file?

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The output I get is as follows:
```
root@4d7382461e21:/workspace/flownet2# python3 main.py
2021-03-19 15:54:31.910652: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic librar
y libcudart.so.11.0
2021-03-19 15:54:33.351618: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_device
s not set
2021-03-19 15:54:33.352532: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic librar
y libcuda.so.1
2021-03-19 15:54:33.438470: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:17:00.0 name: GeForce RTX 2080 Ti computeCapability: 7.5
coreClock: 1.62GHz coreCount: 68 deviceMemorySize: 10.76GiB deviceMemoryBandwidth: 573.69GiB/s
2021-03-19 15:54:33.439233: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 1 with properties:
pciBusID: 0000:65:00.0 name: GeForce RTX 2080 Ti computeCapability: 7.5
coreClock: 1.62GHz coreCount: 68 deviceMemorySize: 10.75GiB deviceMemoryBandwidth: 573.69GiB/s
2021-03-19 15:54:33.439303: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic librar
y libcudart.so.11.0
2021-03-19 15:54:33.439397: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic librar
y libcublas.so.11
2021-03-19 15:54:33.439440: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic librar
y libcublasLt.so.11
2021-03-19 15:54:33.439484: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic librar
y libcufft.so.10
2021-03-19 15:54:33.439527: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2021-03-19 15:54:33.443278: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2021-03-19 15:54:33.444066: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2021-03-19 15:54:33.444126: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-03-19 15:54:33.446050: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0, 1
2021-03-19 15:54:33.446391: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-03-19 15:54:33.447649: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-03-19 15:54:33.614936: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:17:00.0 name: GeForce RTX 2080 Ti computeCapability: 7.5
coreClock: 1.62GHz coreCount: 68 deviceMemorySize: 10.76GiB deviceMemoryBandwidth: 573.69GiB/s
2021-03-19 15:54:33.615494: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 1 with properties:
pciBusID: 0000:65:00.0 name: GeForce RTX 2080 Ti computeCapability: 7.5
coreClock: 1.62GHz coreCount: 68 deviceMemorySize: 10.75GiB deviceMemoryBandwidth: 573.69GiB/s
2021-03-19 15:54:33.615560: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-03-19 15:54:33.615574: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-03-19 15:54:33.615609: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2021-03-19 15:54:33.615623: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2021-03-19 15:54:33.615635: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2021-03-19 15:54:33.615723: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2021-03-19 15:54:33.615762: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2021-03-19 15:54:33.615778: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-03-19 15:54:33.617531: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0, 1
2021-03-19 15:54:33.617607: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-03-19 15:54:34.312786: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-03-19 15:54:34.312840: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0 1
2021-03-19 15:54:34.312848: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N N
2021-03-19 15:54:34.312853: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 1:   N N
2021-03-19 15:54:34.314939: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10069 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2080 Ti, pci bus id: 0000:17:00.0, compute capability: 7.5)
2021-03-19 15:54:34.316283: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 9825 MB memory) -> physical GPU (device: 1, name: GeForce RTX 2080 Ti, pci bus id: 0000:65:00.0, compute capability: 7.5)
2021-03-19 15:54:34.754017: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2021-03-19 15:54:34.772736: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 3299990000 Hz
Epoch 1/480
2021-03-19 15:55:54.178077: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-03-19 15:55:56.009098: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-03-19 15:55:56.406221: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2500/2500 [==============================] - 31685s 13s/step - loss: 965251.6302 - flow2_loss: 2211748.8726 - flow3_loss: 2211866.4072 - flow4_loss: 2213292.5525 - flow5_loss: 2219776.0292 - flow6_loss: 2219457.2435 - flow2_aae: 273742796.5962 - flow3_aae: 273851546.8446 - flow4_aae: 274638881.0683 - flow5_aae: 275303256.6630 - flow6_aae: 275404682.2226 - val_epe: 2226169.7500 - val_aae: 17670296.0000

Epoch 00001: loss improved from inf to 964092.00000, saving model to /workspace/flownet2/milestone1/best/
2021-03-20 00:42:41.302179: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.

Epoch 00001: saving model to /workspace/flownet2/milestone1/last/
Epoch 2/480
2500/2500 [==============================] - 4285s 2s/step - loss: 919111.6756 - flow2_loss: 2113024.2584 - flow3_loss: 2112958.6253 - flow4_loss: 2113410.2004 - flow5_loss: 2113135.0545 - flow6_loss: 2112805.9978 - flow2_aae: 56491300.6014 - flow3_aae: 56474786.1364 - flow4_aae: 56593601.0892 - flow5_aae: 56521067.3974 - flow6_aae: 56437117.0523 - val_epe: 2226564.2500 - val_aae: 17701936.0000

Epoch 00002: loss improved from 964092.00000 to 921045.31250, saving model to /workspace/flownet2/milestone1/best/

Epoch 00002: saving model to /workspace/flownet2/milestone1/last/
{'loss': [964092.0, 921045.3125], 'flow2_loss': [2194216.25, 2117453.25], 'flow3_loss': [2195233.75, 2117392.75], 'flow4_loss': [2199682.75, 2117850.75], 'flow5_loss': [2219717.75, 2117574.25], 'flow6_loss': [2217494.0, 2117253.5], 'flow2_aae': [545963968.0, 59821748.0], 'flow3_aae': [546967424.0, 59804596.0], 'flow4_aae': [550095296.0, 59930716.0], 'flow5_aae': [552018432.0, 59854408.0], 'flow6_aae': [551568128.0, 59766340.0], 'val_epe': [2226169.75, 2226564.25], 'val_aae': [17670296.0, 17701936.0]}
root@4d7382461e21:/workspace/flownet2#
```

**Comment by HungWoeiNeoh** [comment]
I found the problem. Turns out I needed to use the `repeat()` function on the dataset. This line of output was not visible in my code (I was using a tmux session):
```
WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 1200000 batches). You may need to use the repeat() function when building your dataset.
```

**Comment by HungWoeiNeoh** [comment]
From what I can tell, this issue is related to #36539 where a dataset iterator is not recreated at every new epoch
